src/ucp/rndv/rndv_put.c

```diff
@@ -555,3 +517,258 @@ ucp_proto_t ucp_rndv_put_mtype_proto = {
+     * performance will not be optimal. */
+    md_map    = init_params->rkey_config_key->md_map &
+                ep_config->key.reachable_md_map &
+                ~ep_config->rndv.rkey_ptr_dst_mds;
+
+    ucs_for_each_bit(md_index, md_map) {
+        cmpt_index = ucp_ep_config_get_dst_md_cmpt(&ep_config->key, md_index);
+        cmpt_attr  = &context->tl_cmpts[cmpt_index].attr;
+        if (cmpt_attr->flags & UCT_COMPONENT_FLAG_RKEY_PTR) {
+            alloc_md_index = md_index;
+            break;
+        }
+    }
+
+    if (alloc_md_index == UCP_NULL_RESOURCE) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    return ucp_proto_rndv_rkey_ptr_mtype_init_params(init_params,
+                                                     alloc_md_index, frag_size);
+}
+
+static void ucp_proto_rndv_rkey_ptr_mtype_completion(uct_completion_t *uct_comp)
+{
+    ucp_request_t *req = ucs_container_of(uct_comp, ucp_request_t,
+                                          send.state.uct_comp);
+
+    ucp_trace_req(req, "ucp_proto_rndv_rkey_ptr_mtype_completion");
+    ucp_proto_rndv_rkey_destroy(req);
+    ucp_proto_request_zcopy_complete(req, req->send.state.uct_comp.status);
+}
+
+static void
+ucp_proto_rndv_rkey_ptr_mtype_copy_completion(uct_completion_t *uct_comp)
+{
+    ucp_request_t *req = ucs_container_of(uct_comp, ucp_request_t,
+                                          send.state.uct_comp);
+
+    ucp_trace_req(req, "ucp_proto_rndv_rkey_ptr_mtype_copy_completion");
+
+    ucp_proto_completion_init(&req->send.state.uct_comp,
+                              ucp_proto_rndv_rkey_ptr_mtype_completion);
+    ucp_proto_request_set_stage(req, UCP_PROTO_RNDV_PUT_STAGE_ATP);
+    ucp_request_send(req);
+}
+
+static ucs_status_t
+ucp_proto_rndv_rkey_ptr_mtype_copy_progress(uct_pending_req_t *uct_req)
+{
+    ucp_request_t *req    = ucs_container_of(uct_req, ucp_request_t, send.uct);
+    ucp_context_h context = req->send.ep->worker->context;
+    uint64_t remote_address           = req->send.rndv.remote_address;
+    ucs_memory_type_t local_mem_type  = req->send.state.dt_iter.mem_info.type;
+    ucs_memory_type_t remote_mem_type = req->send.rndv.rkey->mem_type;
+    ucp_md_map_t md_map               = req->send.rndv.rkey->md_map;
+    void *rkey_buffer                 = req->send.rndv.rkey_buffer;
+    const ucp_proto_rndv_rkey_ptr_mtype_priv_t
+                               *rpriv = req->send.proto_config->priv;
+    ucp_lane_index_t mem_type_rma_lane;
+    ucp_ep_peer_mem_data_t *ppln_data;
+    ucp_ep_h mem_type_ep;
+    unsigned rkey_index;
+    void *local_ptr;
+    ucs_status_t status;
+
+    ucs_assert(!(req->flags & UCP_REQUEST_FLAG_PROTO_INITIALIZED));
+    ucs_assert(req->send.rndv.rkey_buffer != NULL);
+    ucs_assert(rpriv->alloc_md_index != UCP_NULL_RESOURCE);
+
+    req->send.rndv.rkey_buffer = NULL;
+
+    mem_type_ep = req->send.ep->worker->mem_type_ep[local_mem_type];
+    if (mem_type_ep == NULL) {
```

Comment:
please add `ucs_trace` explaining the error

Reply:
btw, i think it is possible to check during initialization of protocol and discard it if no proper mem_type_ep 
(sorry for the confusion)

Reply:
https://github.com/openucx/ucx/blob/a08930a9d74e9d25dd404a231df6d1240903714f/src/ucp/rndv/rndv_mtype.inl#L21-L26

Given that same check is already in the proto init, it can be removed here